python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+        thread = _WorkerThreadA(self)
+        thread.task_failed.connect(self._on_task_failed)
+        thread.task_completed.connect(self._on_task_completed)
+        self._all_threads.append(thread)
+
+        # start the thread - this will just put it into wait mode:
+        thread.start()
+
+        # log some debug:
+        self._log("Started new background worker thread (num threads=%d)" % len(self._all_threads))
+
+        return thread
+
+    def _start_tasks(self):
+        """
+        Start any queued tasks that are startable if there are available threads to run them.
+        """
+        # start tasks until we fail to start one for whatever reason:
+        started = True
+        while started:
+            started = self._start_next_task()
+
+    def _start_next_task(self):
+        """
+        Start the next task in the queue if there is a task that is startable and there is an 
+        available threads to run it.
+
+        :returns:    True if a task was started, otherwise False
+        """
+        if not self._can_process_tasks:
+            return False
+
+        # figure out next task to start from the priority queue:
+        task_to_process = None
+        task_index = 0
+        priorities = sorted(self._pending_tasks_by_priority.keys(), reverse=True)
+        for priority in priorities:
+            # iterate through the tasks and make sure we aren't waiting on the
+            # completion of any upstream tasks:
+            for ti, task in enumerate(self._pending_tasks_by_priority[priority]):
+                awaiting_upstream_task_completion = False
+                for us_task_id in self._upstream_task_map.get(task.uid, []):
+                    if us_task_id in self._tasks_by_id:
+                        # if the task is still in the tasks list then we're still awaiting
+                        # completion of it!
+                        awaiting_upstream_task_completion = True
+                        break
+                if awaiting_upstream_task_completion:
+                    continue
+
+                # ok, we've found the next task to process:
+                task_to_process = task
+                task_index = ti
+                break
+
+            if task_to_process:
+                # no need to look any further!
+                break
+
+        if not task_to_process:
+            # nothing to do!
+            return False
+
+        # we need a thread to do the work with:
+        thread = self._get_worker_thread()
+        if not thread:
+            # looks like we can't do anything!
+            return False
+
+        self._log("Starting task %r" % task_to_process)
+
+        # ok, we have a thread so lets move the task from the priority queue to the running list:
+        self._pending_tasks_by_priority[priority] = (self._pending_tasks_by_priority[priority][:task_index] 
```

Comment:
hmm, not sure why I wrote that the way I did..!  Not only is it horrid, it's also completely unnecessary as I could just do:

```python
self._pending_tasks_by_priority[priority].remove(task_to_process)
```